### PR TITLE
[Sparse-Gluon] embedding with sparse grad

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
Add sparse_grad option to hybrid embedding block. This block doesn't have optimizations for distributed training. 

Manually checked word_language_model example with sparse_grad=True (without grad clipping). The result is exactly the same as the original implementation. 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: python/mxnet/gluon/trainer.py [hunk 1]
```diff
@@ -110,16 +110,29 @@ def _init_optimizer(self, optimizer, optimizer_params):
                             for _ in self._contexts]
 
     def _init_kvstore(self):
-        arg_arrays = {param.name: param.data(self._contexts[0]) for param in self._params}
+        arg_arrays = {}
+        contains_sparse = False
+        for param in self._params:
+            arg_arrays[param.name] = param.data(self._contexts[0])
```
Comment (szha) [inline python/mxnet/gluon/trainer.py]: did you try to do step on parameter with deferred initialization? what message did you get?

File: python/mxnet/gluon/trainer.py [hunk 2]
```diff
@@ -110,16 +110,29 @@ def _init_optimizer(self, optimizer, optimizer_params):
...
+            if param._grad_stype != 'default':
+                contains_sparse = True
+                # update_on_kvstore is set to False by the user
+                if self._update_on_kvstore is False:
```
Comment (szha) [inline python/mxnet/gluon/trainer.py]: if not self._update_on_kvstore

Comment (eric-haibin-lin) [reply python/mxnet/gluon/trainer.py]: if self._update_on_kvstore is None, I don't need to throw the err

File: python/mxnet/gluon/trainer.py (hunk 1 above)
Comment (eric-haibin-lin) [reply python/mxnet/gluon/trainer.py]: Yes I tried the language model example. param.data() was called in the previous implementation (line 113)

File: python/mxnet/gluon/trainer.py (hunk 2 above)
Comment (szha) [reply python/mxnet/gluon/trainer.py]: OK

File: python/mxnet/gluon/nn/basic_layers.py [hunk 3]
```diff
@@ -390,13 +390,14 @@ class Embedding(HybridBlock):
         - **out**: N-D tensor with shape: `(x1, x2, ..., xN-1, output_dim)`.
     """
     def __init__(self, input_dim, output_dim, dtype='float32',
-                 weight_initializer=None, **kwargs):
+                 weight_initializer=None, sparse_grad=False, **kwargs):
```
Comment (piiswrong) [inline python/mxnet/gluon/nn/basic_layers.py]: documentation for argument

Comment (eric-haibin-lin) [reply python/mxnet/gluon/nn/basic_layers.py]: Added

**Pull request merged**